tonic/src/codec/compression.rs

```diff
@@ -10,62 +10,79 @@ pub(crate) const ENCODING_HEADER: &str = "grpc-encoding";
+    pub fn enable(&mut self, encoding: CompressionEncoding) {
+        for e in self.inner.iter_mut() {
+            match e {
+                Some(e) if *e == encoding => return,
+                None => {
+                    *e = Some(encoding);
+                    return;
+                }
+                _ => continue,
+            }
         }
     }
 
-    /// Enable a [`CompressionEncoding`].
-    pub fn enable(&mut self, encoding: CompressionEncoding) {
-        match encoding {
-            #[cfg(feature = "gzip")]
-            CompressionEncoding::Gzip => self.gzip = true,
-            #[cfg(feature = "zstd")]
-            CompressionEncoding::Zstd => self.zstd = true,
+    /// Remove the last [`CompressionEncoding`].
+    pub fn pop(&mut self) -> Option<CompressionEncoding> {
+        for entry in self.inner.iter_mut().rev() {
+            match entry {
+                Some(e) => {
+                    let removed = *e;
+                    *entry = None;
+                    return Some(removed);
+                }
+                _ => continue,
+            }
         }
+
+        None
     }
 
     pub(crate) fn into_accept_encoding_header_value(self) -> Option<http::HeaderValue> {
-        match (self.is_gzip_enabled(), self.is_zstd_enabled()) {
-            (true, false) => Some(http::HeaderValue::from_static("gzip,identity")),
-            (false, true) => Some(http::HeaderValue::from_static("zstd,identity")),
-            (true, true) => Some(http::HeaderValue::from_static("gzip,zstd,identity")),
-            (false, false) => None,
+        let mut value = BytesMut::new();
+        for encoding in self.inner.iter().copied().filter_map(|e| e) {
+            if !value.is_empty() {
+                value.put_slice(b",");
+            }
+            value.put_slice(encoding.as_str().as_bytes());
         }
-    }
 
-    #[cfg(feature = "gzip")]
-    const fn is_gzip_enabled(&self) -> bool {
-        self.gzip
+        if value.is_empty() {
+            return None;
+        }
+
+        value.put_slice(b",identity");
+        Some(http::HeaderValue::from_maybe_shared(value).unwrap())
     }
 
-    #[cfg(not(feature = "gzip"))]
-    const fn is_gzip_enabled(&self) -> bool {
-        false
+    /// Check if a [`CompressionEncoding`] is enabled.
+    pub fn is_enabled(&self, encoding: CompressionEncoding) -> bool {
+        self.inner.iter().any(|e| *e == Some(encoding))
     }
 
-    #[cfg(feature = "zstd")]
-    const fn is_zstd_enabled(&self) -> bool {
-        self.zstd
+    /// Check if any [`CompressionEncoding`]s are enabled.
+    pub fn is_empty(&self) -> bool {
+        self.inner.iter().all(|e| e.is_none())
     }
+}
 
-    #[cfg(not(feature = "zstd"))]
-    const fn is_zstd_enabled(&self) -> bool {
-        false
+impl Default for EnabledCompressionEncodings {
+    fn default() -> Self {
+        Self {
+            inner: [None, None],
+        }
     }
 }
```

Comment:
This could be implemented with derive macro.

Reply:
Done.